src/psyclone/domain/lfric/kernel/lfric_kernel_metadata.py

```diff
@@ -0,0 +1,506 @@
+            for entry in meta_args:
+                if not isinstance(entry, CommonArg):
+                    raise TypeError(
+                        f"meta_args should be a list of argument objects, "
+                        f"but found {type(entry).__name__}.")
+            self._meta_args = meta_args
+        if meta_funcs is None:
+            self._meta_funcs = []
+            # TBD
+        if meta_reference_element is None:
+            self._meta_reference_element = []
+            # TBD
+        if meta_mesh is None:
+            self._meta_mesh = []
+            # TBD
+
+        self._procedure_name = None
+        if procedure_name is not None:
+            self.procedure_name = procedure_name
+        self._name = None
+        if name is not None:
+            self.name = name
+
+    @staticmethod
+    def create_from_psyir(symbol):
+        '''Create a new instance of LFRicKernelMetadata populated with
+        metadata from a kernel in language-level PSyIR.
+
+        :param symbol: the symbol in which the metadata is stored \
+            in language-level PSyIR.
+        :type symbol: :py:class:`psyclone.psyir.symbols.DataTypeSymbol`
+
+        :returns: an instance of LFRicKernelMetadata.
+        :rtype: :py:class:`psyclone.domain.lfric.kernel.psyir.\
+            LFRicKernelMetadata`
+
+        :raises TypeError: if the symbol argument is not the expected \
+            type.
+        :raises InternalError: if the datatype of the provided symbol \
+            is not the expected type.
+
+        '''
+        if not isinstance(symbol, DataTypeSymbol):
+            raise TypeError(
+                f"Expected a DataTypeSymbol but found a "
+                f"{type(symbol).__name__}.")
+
+        datatype = symbol.datatype
+
+        if not isinstance(datatype, UnknownFortranType):
+            raise InternalError(
+                f"Expected kernel metadata to be stored in the PSyIR as "
+                f"an UnknownFortranType, but found "
+                f"{type(datatype).__name__}.")
+
+        # In an UnknownFortranType, the declaration is stored as a
+        # string, so use create_from_fortran_string()
+        return LFRicKernelMetadata.create_from_fortran_string(
+            datatype.declaration)
+
+    @staticmethod
+    def create_from_fortran_string(fortran_string):
+        '''Create a new instance of LFRicKernelMetadata populated with
+        metadata stored in a fortran string.
+
+        :param str fortran_string: the metadata stored as Fortran.
+
+        :returns: an instance of LFRicKernelMetadata.
+        :rtype: :py:class:`psyclone.domain.lfric.kernel.psyir.\
+            LFRicKernelMetadata`
+
+        :raises ValueError: if the string does not contain a fortran \
+            derived type.
+        :raises ValueError: if kernel metadata is not a Fortran \
+            derived type.
+        :raises ParseError: if meta args is not a list.
```

Comment:
"if the meta_args extracted from the fortran_string is not a list."

Reply:
Modified.

Reply:
already done